File: LeetCode/bag-of-tokens.py
class Solution:
    def bagOfTokensScore(self, tokens: List[int], power: int) -> int:
        tokens.sort()
        l,r,s,m=0,len(tokens)-1,0,0
        while l<=r:
            if power>=tokens[l]:
                power-=tokens[l]
                s+=1
                l+=1
                m=max(s,m)
            elif s>=1 and l<r:
                power+=tokens[r]
                s-=1
                r-=1
            else:
                break
        return m

File: LeetCode/bag-of-tokens.java
class Solution {
    public int bagOfTokensScore(int[] tokens, int power) {
        Arrays.sort(tokens);
        int l=0;
        int h=tokens.length-1;
        int s=0;
        int m=0;
        while (l<=h){
            if(power>=tokens[l]){
                s++;
                power-=tokens[l];
                m=Math.max(m,s);
                l++;
            }
            else if(s>=1){
                s--;
                power+=tokens[h];
                h--;
            }
            else{
                break;
            }
        }
        return m;
    }
}